2021/day17/day17.java

```diff
@@ -0,0 +1,103 @@
+        }
+        return true;
+    }
+
+    public static int calculateHeight(int yVelocity) {
+        int y = 0;
+        while (yVelocity > 0) {
+            y += yVelocity;
+            yVelocity--;
+        }
+        return y;
+    }
+
+    public static boolean isXInTrench(int xVelocity, int xStart, int xEnd) {
+        int x = 0;
+        while (!(x >= xStart && x <= xEnd)) {
+            x += xVelocity;
+            if (x > xEnd || xVelocity == 0) {
+                return false;
+            }
+            xVelocity--;
+        }
+        return true;
+    }
+
+    public static boolean isInTrench(int xVelocity, int yVelocity,
+            int xStart, int xEnd, int yStart, int yEnd) {
+        int x = 0;
+        int y = 0;
+        while (!(x >= xStart && x <= xEnd && y >= yEnd && y <= yStart)) {
+            x += xVelocity;
+            y += yVelocity;
+            if (x < xStart && xVelocity == 0) {
+                return false;
+            } else if (x > xEnd || y < yEnd) {
+                return false;
+            }
+            if (xVelocity > 0) {
+                xVelocity--;
+            }
+            yVelocity--;
+        }
+        return true;
+    }
+
+    public static int[] readInput(String filename){
+        int[] trenchCoordinates = new int[4];
+        try {
+            Scanner sc = new Scanner(new File(filename));
+            String[] newLine = sc.nextLine().split(":")[1].split(",");
+            String[] coordinatesX = newLine[0].split("=")[1].split("\\..");
+            String[] coordinatesY = newLine[1].split("=")[1].split("\\..");
+            trenchCoordinates[0] = Integer.parseInt(coordinatesX[0]);
+            trenchCoordinates[1] = Integer.parseInt(coordinatesX[1]);
+            trenchCoordinates[2] = Integer.parseInt(coordinatesY[1]);
+            trenchCoordinates[3] = Integer.parseInt(coordinatesY[0]);
+            sc.close();
+        } catch (Exception e) {
+            System.out.println("There was an error " + e);
+        }
+        return trenchCoordinates;
+    }
+    public static void main(String[] args) {
+        int[] trenchCoordinates = readInput("data.txt");
+        int xStart = trenchCoordinates[0];
+        int xEnd = trenchCoordinates[1];
+        int yStart = trenchCoordinates[2];
+        int yEnd = trenchCoordinates[3];
+        ArrayList<Integer> xVelocities = new ArrayList<>();
+        ArrayList<Integer> yVelocities = new ArrayList<>();
+        for (int i = -500; i < 500; i++) {
+            if (isXInTrench(i, xStart, xEnd)) {
+                xVelocities.add(i);
+            }
+            if (isYInTrench(i, yStart, yEnd)) {
+                yVelocities.add(i);
+            }
+        }
+        int total = 0;
+        for (Integer xVelocity : xVelocities) {
+            for (Integer yVelocity : yVelocities) {
+                if (isInTrench(xVelocity, yVelocity, xStart, xEnd, yStart, yEnd)) {
+                    total++;
+                }
+            }
+        }
```

Comment:
I guess this is the reason why you have all the functions above even though they're basically all the same logic with minor differences. Maybe you could write just one function with additional arguments `boolean checkX, boolean checkY` that would do the calculations only for the axes that have this `boolean`s set to `true`? I don't exactly know how to make this work and be nice, but I just don't like the repetition.